baseline-network-linux-hld.md

```diff
@@ -0,0 +1,162 @@
+
+Possible APIs are:
+
+- SONiC PDDF, due to the large amount of available platforms
+  https://github.com/sonic-net/SONiC/blob/master/doc/platform/brcm_pdk_pddf.md
+- S3IP as described in https://github.com/sonic-net/SONiC/pull/1068
+- ONLP, as used by ONL and DentOS
+
+The base networking hardware support is organized in multiple repositories:
+
+- kernel: Linux fork with required base changes (backports of important
+submissions, etc)
+- platform-support: drivers, libraries and userspace code for initializing
+platforms
+(do we need more?)
+- TBD: organization of platform support
+- TBD: platform initialization
+
+In the first iteration, the platform support is fine as-is. Future iterations
+should focus on improving the platform drivers and support code to follow modern
+code style and compile with as few warnings as possible, ideally none.
+
+New platform support should follow modern coding styles from the beginning (e.g.
+kernel coding style for modules).
+
+## Design
+
+![example layering with operating layers on top like meta-dent, the proposed
+hardware support layer in the middle, and base Yocto and Open Embedded layers
+at the bottom](images/layering-example.png "layering example")
+
+### Yocto based example and testing OS
+
+To allow quick testing and easy extension for other users, an example
+implementation and support layer for Yocto will be provided. It will contain all
+basic building blocks for creating an installable image:
+
+- recipes for the basic kernel mentioned above and for platform support
+- class and image recipe for an ONIE installable image
+- machine definitions as generic OS-based
+
+### Kernel recipe and configuration handling
+
+The kernel configuration will be done via
+https://docs.yoctoproject.org/kernel-dev/advanced.html to allow large
+flexibility in how other projects using the layer can customize the kernel.
+
+TBD: should we have patches or branches with backport commits? Maybe both?
+
+- patches make required changes due to upstream changes more obvious (provide a
+“conflict diff”) and easier to review
+- branches are closer to the upstream approach, but may be harder to track
+  changes needed due to upstream changes
+
+### Platform support recipes
+
+To allow easier extension by vendors or other users, the platform support will
+be split into multiple subpackages:
+
+- a base package for the userspace utilities and API libraries
+- one or multiple support packages for the platforms
+
+### ASIC/dataplane support
+
+ASIC/dataplane support is out of scope for now, but may be provided by a future
+project or extension.
+
+The only exception are switchdev-enabled platforms like Mellanox or Marvell
+Prestera, where the ASIC support is already provided by the upstream Linux kernel.
+
+### Targeted Kernel and Yocto versions
+
+Target Yocto version is scarthgarp, which is the next LTS release, planned for
+April 2024.
+
+The initial Kernel version target could be 5.15 (minimum kernel version of Yocto)
```

Comment:
"will be 5.15"?

Reply:
I think we probably will need to update this once we all agree on the details here and can pinpoint the best target kernel a bit better.

Reply:
Right, my idea is to start out the lowest supported kernel, since will pose the least amount of work to do to get someting running, but then quickly switch to a more recent LTS, or maybe even the most recent.

Which one we target is probably something that will move if we do try to become an official Yocto layer.